File: spec/feature/feature_spec.rb
require 'airport'
require 'plane'

RSpec.describe 'all objects and methods' do
  let(:small_airport) { Airport.new(1) }
  let(:stormy_airport) { Airport.new }
  let(:plane1) { Plane.new }
  let(:plane2) { Plane.new }
  let(:plane3) { Plane.new }
  before(:each) do
    # Allow sunny weather and land plane for take off test
    allow(small_airport).to receive(:stormy?) { nil }
    small_airport.land(plane1)
    # Allow sunny weather, land plane for take off test then make this airport stormy
    allow(stormy_airport).to receive(:stormy?) { nil }
    stormy_airport.land(plane3)
    allow(stormy_airport).to receive(:stormy?) { true }
  end
  it 'feature test' do
    expect(small_airport.in_hanger?(plane1)).to be true
    expect { small_airport.land(plane3) }.to raise_error("Plane is not in air")
    expect { small_airport.land(plane1) }.to raise_error("Plane has already landed")
    expect { small_airport.land(plane2) }.to raise_error("Airport is full")
    expect(small_airport.take_off(plane1)).to be plane1
    expect(small_airport.in_hanger?(plane1)).to be false
    expect { small_airport.take_off(plane1) }.to raise_error("Plane not in hanger")
    expect { stormy_airport.land(plane2) }.to raise_error("Too stormy for landing")
    expect { stormy_airport.take_off(plane3) }.to raise_error("Too stormy for take off")
  end
end

File: jasmine/spec/featureSpec.js
'use strict';

describe('Feature Test: ', function(){
  var plane;
  var airport;

  beforeEach(function(){
    plane = new Plane();
    airport = new Airport();
  });
  describe('under normal conditions',function(){
    beforeEach(function(){
      spyOn(Math,'random').and.returnValue(0);
    });
    // As an air traffic controller
    // To get passengers to a destination
    // I want to instruct a plan to land at
    //     an airport and confirm that it has landed
    it('planes can be instructed to land at an airport', function(){
      plane.land(airport);
      expect(airport.planes()).toContain(plane);
    });
    // As an air traffic controller
    // To get passengers to a destination
    // I want to instruct a plan to take off from
    //     an airport and confirm that it is no longer in the airport
    it('planes can be instructed to takeoff', function(){
      plane.land(airport)
      plane.takeoff();
      expect(airport.planes()).not.toContain(plane);
    });
  });
  describe('under stormy conditions',function(){
    // As an air traffic controller
    // To ensure safety
    // I want to prevent takeoff when weather is stormy
    it('blocks takeoff when weather is stormy', function(){
      spyOn(Math,'random').and.returnValue(0);
      plane.land(airport)
      spyOn(airport._weather,'isStormy').and.returnValue(true);
      expect(function(){ plane.takeoff();}).toThrowError('cannot takeoff during storm');
      expect(airport.planes()).toContain(plane);
    });
    // As an air traffic controller
    // To ensure safety
    // I want to prevent landing when weather is stormy
    it('blocks landing when weather is stormy', function(){
      spyOn(Math,'random').and.returnValue(1);
      expect(function(){ plane.land(airport); }).toThrowError('cannot land during storm');
      expect(airport.planes()).toEqual([]);
    });
  });
});
